pydomo/Transport.py

```diff
@@ -71,23 +73,41 @@ def request(self, url, method, headers, params=None, body=None):
                         'params': params, 'data': body, 'stream': True}
         if self.request_timeout:
             request_args['timeout'] = self.request_timeout
-        response = requests.request(**request_args)
-        if response.status_code == requests.codes.UNAUTHORIZED:
+
+        # Expiration date should be in UTC
+        if datetime.now(timezone.utc).timestamp() > self.token_expiration:
+            self.logger.debug("Access token is expired")
             self._renew_access_token()
             headers['Authorization'] = 'bearer ' + self.access_token
-            response = requests.request(**request_args)
-        return response
+
+        return requests.request(**request_args)
 
     def _renew_access_token(self):
         self.logger.debug("Renewing Access Token")
         url = self.apiHost + '/oauth/token?grant_type=client_credentials'
         response = requests.post(url=url, auth=HTTPBasicAuth(self.clientId, self.clientSecret))
         if response.status_code == requests.codes.OK:
             self.access_token = response.json()['access_token']
+            self.token_expiration = self._extract_expiration(self.access_token)
         else:
             self.logger.debug('Error retrieving access token: ' + self.dump_response(response))
             raise Exception("Error retrieving a Domo API Access Token: " + response.text)
 
+    def _extract_expiration(self, access_token):
+        EXPIRATION_DATE_KEY = 'exp'
+
+        token_parts = access_token.split('.')
+        payload_bytes = bytes(token_parts[1], 'utf-8')
```

Comment:
How do we handle if we have trouble parsing the token? Should we be catching all errors and returning unauthorized if there's a token error?